### PR TITLE
[DQL-183] Follow activity enhancements

### PR DESCRIPTION
**Pull request by salsa-nathan**
This PR contains changes for the following stories:

- [DQL2-18] Follow activity email notification enhancements
- [DQL2-20] Automated tests
- [DQL2-21] Follow activity login rules

@duttonw & @ThrawnCA for your review.

### CODE REVIEW

File: ckanext/data_qld/controller.py [hunk 1]
```diff
@@ -74,3 +79,51 @@ def show_schema(self, dataset_id, resource_id):
...
             tk.abort(403, tk._('You are not authorized to view the Data Scheme for the resource %s' % resource_id))
+
+
+class DataQldDataset(PackageController):
+    def _get_context(self):
+        return {'model': model, 'session': model.Session,
+                       'user': c.user, 'for_view': True,
+                       'auth_user_obj': c.userobj}
+
+    def _is_dataset_private(self, id):
```
[the remark this answers is not in the document]
Comment (salsa-nathan) [reply ckanext/data_qld/controller.py:88]: @ThrawnCA please see comment added to https://salsadigital.atlassian.net/browse/DQL2-20

File: test/features/login_redirect.feature [hunk 2]
```diff
@@ -0,0 +1,55 @@
+@login_redirect
+Feature: Login Redirection
+
+    @dashboard_login
+    Scenario: As an unauthenticated user, when I visit the /dashboard URL I see the login page
+        Given "Unathenticated" as the persona
```
[the remark this answers is not in the document]
Comment (salsa-nathan) [reply test/features/login_redirect.feature]: @ThrawnCA Updated username "Unauthenticated" in tests after updating typo in `environment.py`

File: ckanext/data_qld/plugin.py [hunk 3]
```diff
@@ -107,6 +112,26 @@ def before_map(self, m):
                   controller='ckanext.data_qld.controller:DataQldUI',
                   action='show_schema', conditions=dict(method=['GET']))
 
+        # This is a pain, but re-assigning the dataset_read route using `before_map` appears to affect these two routes,
```
[the remark this answers is not in the document]
Comment (salsa-nathan) [reply ckanext/data_qld/plugin.py]: @ThrawnCA Adjusted length of code comment in `plugin.py`

File: test/features/login_redirect.feature [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+    Scenario: As an unauthenticated organisation member, when I visit the URL of a private dataset I see the login page. Upon logging in I am taken to the private dataset
+        Given "TestOrgMember" as the persona
+        When I visit "/dataset/annakarenina"
+        Then I should see an element with xpath "//h1[contains(string(), 'Login')]"
+        When I log in directly
+        Then I should see an element with xpath "//h1[contains(string(), 'A Novel By Tolstoy')]"
+        And I should see an element with xpath "//span[contains(string(), 'Private')]"
+
+    @private_dataset_resource
+    Scenario: As an unauthenticated organisation member, when I visit the URL of a private dataset resource I see the login page. Upon logging in I am taken to the private dataset resource
```
[the remark this answers is not in the document]
Comment (salsa-nathan) [reply test/features/login_redirect.feature]: @ThrawnCA Added BDD test for logged in org member accessing dataset private to org

File: test/features/login_redirect.feature [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+        Given "Admin" as the persona
+        When I log in
+        Then I visit "/dataset/edit/annakarenina"
+        When I select "True" from "private"
+        And I fill in "author_email" with "admin@localhost"
+        And I press "Update Dataset"
+        Then I should see an element with xpath "//span[contains(string(), 'Private')]"
+
+    @private_dataset
+    Scenario: As an unauthenticated user, when I visit the URL of a private dataset I see the login page
```
[the remark this answers is not in the document]
Comment (salsa-nathan) [reply test/features/login_redirect.feature:35]: @ThrawnCA Added BDD test for unauthenticated user visiting a public dataset

File: ckanext/data_qld/controller.py [hunk 6]
```diff
@@ -74,3 +79,49 @@ def show_schema(self, dataset_id, resource_id):
...
+    def resource_read(self, id, resource_id):
+        """
+        Override the default CKAN behaviour for private Dataset Resource visibility for unauthenticated users.
+        Instead of displaying "404 Dataset not found" message, give unauthenticated users a chance to login (if the
+        dataset exists but is private).
+        :param id: Package id/name
+        :param resource_id: Resource id
+        :return:
+        """
+        if not g.user and not self._is_dataset_private(id):
```
Comment (ThrawnCA) [inline ckanext/data_qld/controller.py:121]: This works, but the name of the function is inverted. It should redirect when the target dataset _is_ private.

**Comment by salsa-nathan** [comment]
@duttonw & @ThrawnCA Hi guys - is this ready to be merged for QA now? thanks.

File: ckanext/data_qld/controller.py [hunk 7]
```diff
@@ -85,7 +85,7 @@ class DataQldDataset(PackageController):
     def _get_context(self):
         return {'model': model, 'session': model.Session, 'user': c.user, 'for_view': True, 'auth_user_obj': c.userobj}
 
-    def _is_dataset_private(self, id):
+    def _is_dataset_public(self, id):
```
Comment (ThrawnCA) [inline ckanext/data_qld/controller.py:88]: It's more about "visible" than "public" (ie either it's public, or we're logged in with sufficient privileges), but ok.

**Comment by salsa-nathan** [comment]
@duttonw & @ThrawnCA - can this be merged into develop now?

**Pull request merged**